File: packages/bdb/src/lib/merkle-tree-ref.ts
import { upsertItem, deleteItem, type IMerkleTree, type HashedItem, buildMerkleTree } from "merkle-tree";
import type { IUuidGenerator } from "utils";

//
// Lazily-loaded, committable handle for a merkle tree.
// Works at shard, collection, and database level via injected callbacks.
//
export interface IMerkleRef<T = undefined> {
    //
    // Returns the underlying merkle tree, invoking the loader on first access.
    // Returns undefined when there is no tree (empty shard/collection/database).
    //
    get(): Promise<IMerkleTree<T> | undefined>;

    //
    // Inserts or updates a hashed item in the tree, creating the tree if needed.
    //
    upsert(item: HashedItem): Promise<void>;

    //
    // Removes an item by name. Sets tree to undefined if it becomes empty.
    //
    remove(name: string): Promise<void>;

    //
    // Persists the tree to storage (or deletes it if empty). No-op when not dirty.
    //
    commit(): Promise<void>;

    //
    // Drops the cached tree from memory so the next access reloads from storage.
    // Throws if dirty — call commit() first.
    //
    flush(): void;
}

//
// Concrete lazily-loaded merkle ref.
// The caller supplies load/save/delete callbacks appropriate for the level
// (shard, collection, or database). An optional creator callback is needed
// when upsert() may be called on a level whose tree starts empty (e.g. database).
//
export class MerkleRef<T = undefined> implements IMerkleRef<T> {
    //
    // Cached tree; undefined when empty or not yet loaded.
    //
    private _tree: IMerkleTree<T> | undefined = undefined;

    //
    // True once get() has completed at least once.
    //
    private _loaded = false;

    //
    // True when the in-memory tree differs from what is on disk.
    //
    private _dirty = false;

    constructor(
        //
        // Loads (and, if needed, builds) the tree. Returns undefined when there is no tree.
        //
        private readonly loader: () => Promise<IMerkleTree<T> | undefined>,

        //
        // Saves the tree to storage.
        //
        private readonly saver: (tree: IMerkleTree<T>) => Promise<void>,

        //
        // Deletes the tree from storage.
        //
        private readonly deleter: () => Promise<void>,

        //
        // Creates a new empty tree when upsert() is called and the tree is undefined.
        // Required when the tree may not exist yet (e.g. first collection in a database).
        //
        private readonly creator: () => Promise<IMerkleTree<T>>,
    ) {
    }

    //
    // Returns the tree, invoking the loader on first access.
    //
    async get(): Promise<IMerkleTree<T> | undefined> {
        if (!this._loaded) {
            this._tree = await this.loader();
            this._loaded = true;
        }

        if (this._tree?.dirty) {
            this._tree.merkle = buildMerkleTree(this._tree.sort);
            this._tree.dirty = false;
        }

        return this._tree;
    }

    //
    // Inserts or updates a hashed item in the tree.
    // If the tree is undefined and a creator was supplied, creates a new tree first.
    //
    async upsert(item: HashedItem): Promise<void> {
        let tree = await this.get();
        if (!tree) {
            tree = await this.creator();
        }

        this._tree = upsertItem(tree, item);
        this._dirty = true;
    }

    //
    // Removes an item by name. Sets the tree to undefined if it becomes empty.
    //
    async remove(name: string): Promise<void> {
        const tree = await this.get();
        if (!tree || !tree.sort) {
            return;
        }

        deleteItem(tree, name);

        if (!tree.sort) {
            this._tree = undefined;
        }

        this._dirty = true;
    }

    //
    // Writes the tree to storage (or deletes it if empty); clears the dirty flag.
    //
    async commit(): Promise<void> {
        if (!this._dirty) {
            return;
        }

        if (this._tree) {
            await this.saver(this._tree);
        }
        else {
            await this.deleter();
        }

        this._dirty = false;
    }

    //
    // Drops the cached tree. Throws if dirty — call commit() first.
    //
    flush(): void {
        if (this._dirty) {
            throw new Error('Cannot flush a dirty MerkleRef — call commit() first.');
        }

        this._tree = undefined;
        this._loaded = false;
    }
}

File: packages/bdb-zig/src/lib/merkle-tree-ref.zig
const std = @import("std");
const merkle_tree_zig = @import("merkle-tree-zig");
const merkle_tree = merkle_tree_zig.merkle_tree;
const IMerkleTree = merkle_tree.IMerkleTree;
const HashedItem = merkle_tree.HashedItem;

//
// Lazily-loaded, committable handle for a merkle tree.
// Works at shard, collection, and database level via injected callbacks.
// (Zig: IMerkleRef has a single implementation, so callers use *MerkleRef directly.)
//
pub const IMerkleRef = MerkleRef;

//
// Loads (and, if needed, builds) the tree. Returns null when there is no tree.
//
pub const Loader = *const fn (context: *anyopaque, allocator: std.mem.Allocator, io: std.Io) anyerror!?IMerkleTree;

//
// Saves the tree to storage.
//
pub const Saver = *const fn (context: *anyopaque, allocator: std.mem.Allocator, io: std.Io, tree: *IMerkleTree) anyerror!void;

//
// Deletes the tree from storage.
//
pub const Deleter = *const fn (context: *anyopaque, allocator: std.mem.Allocator, io: std.Io) anyerror!void;

//
// Creates a new empty tree when upsert() is called and the tree is undefined.
//
pub const Creator = *const fn (context: *anyopaque, allocator: std.mem.Allocator, io: std.Io) anyerror!IMerkleTree;

//
// Concrete lazily-loaded merkle ref.
// The caller supplies load/save/delete callbacks appropriate for the level
// (shard, collection, or database). An optional creator callback is needed
// when upsert() may be called on a level whose tree starts empty (e.g. database).
// (Zig: the TypeScript arrow functions become function pointers that receive `context`, the object that owns the
// ref. The ref keeps the allocator it was created with; tree nodes are allocated with it.)
//
pub const MerkleRef = struct {
    //
    // Allocates the tree nodes (must live as long as the ref).
    //
    allocator: std.mem.Allocator,

    //
    // Cached tree; null when empty or not yet loaded.
    //
    _tree: ?IMerkleTree = null,

    //
    // True once get() has completed at least once.
    //
    _loaded: bool = false,

    //
    // True when the in-memory tree differs from what is on disk.
    //
    _dirty: bool = false,

    //
    // The object the callbacks act on (the shard, collection or database).
    //
    context: *anyopaque,

    //
    // Loads (and, if needed, builds) the tree. Returns undefined when there is no tree.
    //
    loader: Loader,

    //
    // Saves the tree to storage.
    //
    saver: Saver,

    //
    // Deletes the tree from storage.
    //
    deleter: Deleter,

    //
    // Creates a new empty tree when upsert() is called and the tree is undefined.
    // Required when the tree may not exist yet (e.g. first collection in a database).
    //
    creator: Creator,

    //
    // Creates a ref (TypeScript: the constructor).
    //
    pub fn init(allocator: std.mem.Allocator, context: *anyopaque, loader: Loader, saver: Saver, deleter: Deleter, creator: Creator) MerkleRef {
        return .{
            .allocator = allocator,
            .context = context,
            .loader = loader,
            .saver = saver,
            .deleter = deleter,
            .creator = creator,
        };
    }

    //
    // Returns the tree, invoking the loader on first access.
    //
    pub fn get(self: *MerkleRef, io: std.Io) !?*IMerkleTree {
        if (!self._loaded) {
            self._tree = try self.loader(self.context, self.allocator, io);
            self._loaded = true;
        }

        if (self._tree) |*tree| {
            if (tree.dirty) {
                tree.merkle = try merkle_tree.buildMerkleTree(self.allocator, tree.sort);
                tree.dirty = false;
            }
            return tree;
        }

        return null;
    }

    //
    // Inserts or updates a hashed item in the tree.
    // If the tree is undefined and a creator was supplied, creates a new tree first.
    //
    pub fn upsert(self: *MerkleRef, io: std.Io, item: HashedItem) !void {
        var tree: IMerkleTree = undefined;
        if (try self.get(io)) |existing| {
            tree = existing.*;
        }
        else {
            tree = try self.creator(self.context, self.allocator, io);
        }

        self._tree = try merkle_tree.upsertItem(self.allocator, &tree, item);
        self._dirty = true;
    }

    //
    // Removes an item by name. Sets the tree to undefined if it becomes empty.
    //
    pub fn remove(self: *MerkleRef, io: std.Io, name: []const u8) !void {
        const tree = try self.get(io) orelse {
            return;
        };
        if (tree.sort == null) {
            return;
        }

        try merkle_tree.deleteItem(self.allocator, tree, name);

        if (tree.sort == null) {
            self._tree = null;
        }

        self._dirty = true;
    }

    //
    // Writes the tree to storage (or deletes it if empty); clears the dirty flag.
    //
    pub fn commit(self: *MerkleRef, io: std.Io) !void {
        if (!self._dirty) {
            return;
        }

        if (self._tree) |*tree| {
            try self.saver(self.context, self.allocator, io, tree);
        }
        else {
            try self.deleter(self.context, self.allocator, io);
        }

        self._dirty = false;
    }

    // Not ported: flush (psi replicate and psi verify never flush a database).
};
